pkg/app/run.go

```diff
@@ -205,5 +219,59 @@ func (r *Run) diff(triggerCleanupEvent bool, detailedExitCode bool, c DiffConfig
 %s
 `, strings.Join(names, "\n"))
 
-	return &infoMsg, releasesToBeUpdated, releasesToBeDeleted, nil
+	return &infoMsg, releasesToBeUpdated, releasesToBeDeleted, releasesToBeRolledback, nil
+}
+
+// Rollback should be performed if the following conditions are met
+//   - --allow-rollback flag has been set
+//   - chart version is provided in the helmfile
+//   - chart version specified in the helmfile is less than the chart version of the latest release revision
+//   - a previous release revision is found with the same chart version as in the helmfile
+//
+// The revision rolled back to shall be the highest revision with the same chart version as provided in the helmfile
+func (r *Run) checkIfShouldRollback(release state.ReleaseSpec, diffOpts *state.DiffOpts) (bool, string) {
+	if !diffOpts.AllowRollback || release.Version == "" {
+		return false, ""
+	}
+	st := r.state
+	helm := r.helm
+
+	deployedVersion, err := st.GetDeployedVersion(helmexec.HelmContext{
+		WorkerIndex: 0,
+		HistoryMax:  1,
+	}, helm, &release)
+	if err != nil {
+		return false, ""
+	}
+
+	lessThanDeployedVersionConstraint, err := semver.NewConstraint("< " + deployedVersion)
+	if err != nil {
+		r.logger.Errorf("error creating deployed version constraint for chart %s, rollback will not be considered, update will be executed: %v", release.Chart, err)
+		return false, ""
+	}
+	versionFromHelmFile, err := semver.NewVersion(release.Version)
+	if err != nil {
+		r.logger.Errorf("error parsing version %s for chart %s, rollback will not be considered, update will be executed: %v", release.Version, release.Chart, err)
+		return false, ""
+	}
+
+	if lessThanDeployedVersionConstraint.Check(versionFromHelmFile) {
+		r.logger.Debugf("chart version is less then deployed chart version for chart: %s", release.Chart)
+		revisionToRollbackTo, err := st.GetHighestRevisionWithVersion(helmexec.HelmContext{
+			WorkerIndex: 0,
+			HistoryMax:  1,
+		}, helm, &release)
+		if err != nil {
+			r.logger.Errorf("error getting highest revision of chart %s with version %s, rollback will not be considered, update will be executed: %v", release.Chart, release.Version, err)
+			return false, ""
+		}
+		if revisionToRollbackTo == "" {
+			r.logger.Debugf("no release revision found for chart %s with version %s, rollback not possible, update will be executed", release.Chart, release.Version)
+			return false, ""
+		}
+		r.logger.Debugf("rollback of chart %s to release %s will be executed", release.Chart, release.Version)
+		return true, revisionToRollbackTo
```

Comment:
@MichaelMorrisEst Hey! Thank you very much for your contribution!

Just started reviewing this. Very interesting.

So, let's say you have a release defined in helmfile.yaml for the chart C and version X, and you previously installed X and upgraded it to X'...

You use helm-rollback to downgrade the release's chart version from X' to X, while reusing the values you previously used when installing the version X, hence ignoring the values defined in helmfile.yaml, right?

Reply:
Hi @mumoshu 

Yes, that is correct, the release will be downgraded from X' to X with helm using the values it has stored in history for X (as per helm rollback functionality)

A subsequent sync/apply would bring the values in line with the values in the helmfile.yaml. Perhaps this new functionality should automatically invoke apply after the rollback to ensure the values are consistent with the helmfile.yaml in order to ensure it does not violate declarative logic?

I'm open to any suggestions on how what I've proposed can be improved

Reply:
@MichaelMorrisEst Thanks for repying!

So helmfile being a declarative deployment tool, I'd expect any helmfile-apply to sync the current state with the desired state defined in the yaml file, not leaving the current state divergent from the desired.

Doing helm-rollback only does seem to leave the current state diverged and that could be no option.
helm-rollback followed by helm-upgrade to sync values- that might be a bit better. However, I'm still concerned that it will result in redundant pod replacements.

That said, I think we have two options:

- Change your chart so that it does everything it does in pre and post rollback hooks in pre/post upgrade hooks instead. This way, we don't need this patch- any helmfile-apply can be either to upgrade or to downgrade. In any case pre/post hooks does its job depending on if it's actually an upgrade or downgrade.
- Change helmfile so that it automatically runs helm-rollback on failed releases on helmfile-apply failure. This does leave diverged current state but, as a failing helmfile-apply means the desired state is invalid and not applicable, it's ok.

I prefer the first option as it would work more universally.


Also, would you mind sharing an example helmfile.yaml and charts that helm-rollback turn useful? That would help deciding what we want helmfile to do!